### ISSUE TITLE
Services with externalIPs Should Honor externalTrafficPolicy

### ISSUE THREAD

**Issue by seils** [kubernetes/kubernetes#69811]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:
Services that define IP addresses using externalIPs (vs. loadBalancerIP) do not adhere to the localTrafficPolicy=Local setting.  This prevents preservation of (client) source IP addresses for such services.

**What you expected to happen**:
Services with externalIPs definitions should respect the localTrafficPolicy setting.

**How to reproduce it (as minimally and precisely as possible)**:
- Create a deployment.
- Expose said deployment via Service with type=LoadBalancer, externalIPs, and localTrafficPolicy=Local.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.7", GitCommit:"0c38c362511b20a098d7cd855f1314dad92c2780", GitTreeState:"clean", BuildDate:"2018-08-20T10:09:03Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.7-gke.6", GitCommit:"06898a4d0f2b96f82b43d9e59fa2825bd3d616a2", GitTreeState:"clean", BuildDate:"2018-10-02T17:32:01Z", GoVersion:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
GCP

- OS (e.g. from /etc/os-release):
```
BUILD_ID=10895.52.0
NAME="Container-Optimized OS"
KERNEL_COMMIT_ID=96d6d803ccae172c7a1e9b0f1fee8bcff8894756
GOOGLE_CRASH_ID=Lakitu
VERSION_ID=69
BUG_REPORT_URL="https://cloud.google.com/container-optimized-os/docs/resources/support-policy#contact_us"
PRETTY_NAME="Container-Optimized OS from Google"
VERSION=69
GOOGLE_METRICS_PRODUCT_ID=26
HOME_URL="https://cloud.google.com/container-optimized-os/docs"
ID=cos
```

- Kernel (e.g. `uname -a`):
```
Linux gke-cluster-1-default-pool-471503ef-ckl6 4.14.65+ #1 SMP Sun Sep 9 02:18:33 PDT 2018 x86_64 Intel(R) Xeon(R) CPU @ 2.60GHz GenuineIntel GNU/Linux
```

- Install tools:
- Others:

**Comment by uablrek** [comment]
The `loadBalancerIP` and the `externalIPs` should be handled in *exactly* the same way by the `kube-proxy`. The are similar from a traffic point of view, just set in different ways. That would probably also simplify the code.

**Comment by uablrek** [comment]
Still valid

**Comment by ahg-g** [comment]
Likely we can fix this by changing the target chain fron KUBE-SVC-* to KUBE-XLB-* for external ip rules. Specifically, when svcInfo.OnlyNodeLocalEndpoints is true, change svcChain to svcXlbChain in  https://github.com/kubernetes/kubernetes/blob/0e224ad3f38d9ab0ffa34bcd0774458f5834b7fa/pkg/proxy/iptables/proxier.go#L937

We also may want to skip adding the masquerade rule for external ips when svcInfo.OnlyNodeLocalEndpoints is true: https://github.com/kubernetes/kubernetes/blob/0e224ad3f38d9ab0ffa34bcd0774458f5834b7fa/pkg/proxy/iptables/proxier.go#L928

**Comment by thockin** [comment]
The main problem I see with this is that `externalIPs` are totally unmanaged.  To make `Local` work (at least on some platforms) we need to fail an LB health-check on any node that does NOT have a backend.  We support that through kube-proxy, but not for externalIPs - we key it off of "type=LoadBalancer".

I suppose we could key it off externalIPs, too.

What I don't know is why this is preferable to using type=LB?  The backends for the "manual" LBs are not kept synchronized ("totally unamanaged"), so as nodes come and go the end-user is responsible for updating the config.  For platforms with managed pools of nodes (ASGs, MIGs, etc) this is fine as long as the set of pools does not change.

In short, externalIPs is one of my very least favorite features, and I feel obligated to talk people out of using them whenever possible.  That said, there are some legit reasons and scenarios where it's the only good answer.

**Comment by seils** [comment]
99% of the time I recommend externalIPs it is to get around the feature gap
between the cloud-native load balancer and what can be instantiated through
the Kubernetes manifest.

**Comment by benceszikora** [comment]
We are running an on-prem cluster and our main use case for `externalIPs` is our ingress VIP. Our ingress service is a NodePort service with `externalTrafficPolicy: Local`. Because this setting is not respected and the packets are masqueraded when using an `externalIP`, we lose the users real IP address from the `X-Forwarded-For` header. At the moment I don't see any workarounds for this either, but if anyone has any suggestions that would be welcome.

We have been using kube-router in our previous clusters (so no kube-proxy), but this became an issue in our calico + kube-proxy setup.

**Comment by thockin** [comment]
Can you use type=LoadBalancer and set the `status.loadBalancer.ingress.ip`
instead?  Normally that is filled in by a cloud provider.  You are,
effectively, your own cloud provider.

**Comment by uablrek** [comment]
If needed please see; https://github.com/Nordix/assign-lb-ip

**Comment by benceszikora** [comment]
I gave that a go, but unfortunately I can't access our ingress at all with that setup. I'll need to do a bit more digging to figure out what's missing exactly.

**Comment by benceszikora** [comment]
@thockin The problem is that calico doesn't advertise `status.loadBalancer.ingress.ip`. It does advertise our entire external service IP range (by default), but not a `/32` for the nodes that the corresponding pods are running on.
This results in requests getting round robined across all the nodes, not just the ones having the service on them. When the requests hit a node that the service *is* running on, things are fine and `X-Forwarded-For` is correct. When is doesn't it fails completely with `No route to host`.
Once I add `externalIPs` it works correctly again, but `X-Forwarded-For` is incorrect.

**Comment by benceszikora** [comment]
This entire setup also works fine with using the cluster `clusterIP` of  the service (as described here https://www.projectcalico.org/kubernetes-service-ip-route-advertisement/), but kube-proxy seems to treat `externalIPs` differently.

**Comment by thockin** [comment]
I am confused now. Are you trying to use a VIP or a cluster IP?  Calico should not be advertising anything for a VIP AFAIK

**Comment by benceszikora** [comment]
Sorry for confusing things. I was just mentioning that hitting the service on its `clusterIP` works fine with `X-Forwarded-For` (provided that `clusterCIDR` is not set in the kube-proxy config, due to this: https://github.com/kubernetes/kubernetes/blob/0e224ad3f38d9ab0ffa34bcd0774458f5834b7fa/pkg/proxy/iptables/proxier.go#L862-L869).

We use the `externalIPs` on our ingress service to get traffic into the cluster. Calico does advertise `externalIPs`: https://docs.projectcalico.org/v3.11/networking/advertise-service-ips#advertise-service-external-ip-addresses

**Comment by benceszikora** [comment]
FWIW, we have rolled out a fork of `kube-proxy` last weekend with @ahg-g's suggested patch and that seems to work perfectly. Are there any downsides to those changes?

**Comment by aojea** [comment]
/close

Fixed by https://github.com/kubernetes/kubernetes/pull/88786
